linkerd/app/core/src/control.rs

```diff
@@ -145,160 +145,376 @@ pub mod add_origin {
+            if !old_targets.contains(new_target) {
+                adds.push((new_target.addr, new_target.clone()))
+            }
         }
 
-        fn call(&mut self, target: ControlAddr) -> Self::Future {
-            let state = match target.addr {
-                Addr::Socket(sa) => State::make_inner(sa, &target, &mut self.inner),
-                Addr::Name(ref na) => {
-                    // The inner service is ready, but we are going to do
-                    // additional work before using it. In case the inner
-                    // service has acquired resources (like a lock), we
-                    // relinquish our claim on the service by replacing it.
-                    self.inner = self.inner.clone();
-
-                    let future = self.dns.resolve_one_ip(na.name());
-                    State::Resolve(future, Some((self.inner.clone(), target.clone())))
-                }
-            };
-
-            Init { state }
+        for old_target in old_targets.iter() {
+            if !new_targets.contains(old_target) {
+                removes.push(old_target.addr)
+            }
         }
-    }
 
-    // === impl Init ===
+        (adds, removes)
+    }
 
-    impl<M> Future for Init<M>
+    pub fn resolution_stream<R>(
+        dns: R,
+        address: ControlAddr,
+    ) -> impl Stream<Item = Result<Update<Target>, Error>>
     where
-        M: tower::Service<client::Target>,
+        R: dns::Resolver,
     {
-        type Output = Result<M::Response, Error<M::Error>>;
-
-        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-            let mut this = self.project();
-            loop {
-                match this.state.as_mut().project() {
-                    StateProj::Resolve(fut, stack) => {
-                        let ip = ready!(fut.poll(cx).map_err(Error::Dns))?;
-                        let (svc, config) = stack.take().unwrap();
-                        let addr = SocketAddr::from((ip, config.addr.port()));
-                        this.state
-                            .as_mut()
-                            .set(State::NotReady(svc, Some((addr, config))));
+        let port = address.addr.port().clone();
+        let server_name = address.identity.clone();
+        try_stream! {
+            match address.clone().addr {
+                Addr::Socket(sa) => {
+                    // this should yield the socket address once
+                    // and keep on yielding NotReady forever.
+                    let mut stream = futures::stream::once(future::ok::<Update<Target>, Error>(
+                        Update::Add(vec![(sa, Target::new(sa, address.identity))]),
+                    ))
+                    .chain(tokio::stream::pending());
+
+                    while let Some(value) = stream.next().await {
+                        debug!(?value, "resolved once");
+                        yield value.expect("no errors in this stream");
```

Comment:
I think this could be a bit simpler:
```suggestion
                    // this should yield the socket address once
                    // and keep on yielding NotReady forever.
                    debug!(?value, "resolved once");
                    yield Ok(Update::Add(vec![(sa, Target::new(sa, address.identity))]));
                    futures::future::pending().await;
```